src/turing_machine/machine.py

```diff
@@ -1,41 +1,83 @@
-class TuringMachine:
-    def __init__(self, tape, initial_state, final_states, transition_function):
+from abc import ABC, abstractmethod
+
+class Tape:
+    def __init__(self, tape):
         self.tape = list(tape)
         self.head = 0
-        self.state = initial_state
+
+    def move_left(self):
+        self.head -= 1
+
+    def move_right(self):
+        self.head += 1
+
+    def get_value(self):
+        return self.tape[self.head]
+
+    def set_value(self, val):
+        self.tape[self.head] = val
+
+    def __str__(self):
+        return ''.join(self.tape)
+
+
+class TuringMachine(ABC):
+    def __init__(self, tapes, initial_state, final_states):
+        self.initial_tapes = [Tape(tape) for tape in tapes]
+        self.initial_state = initial_state
         self.final_states = final_states
-        self.transition_function = transition_function
+        self.reset()
+
+    def reset(self):
+        self.tapes = [Tape(tape.tape) for tape in self.initial_tapes]
```

Comment:
Tutaj wystąpi problem przekazania wartości przez referencję. Czyli w tym przypadku, składowa `self.tapes` będzie przechowywać referencje do listy `self.initial_tapes`. W ten sposób, zmieniając wartość dla jednej z list w `self.tapes` będziemy jednocześnie zmieniać wartość dla `self.initial_tapes`. Dlatego zrobiłbym to w trochę inny sposób.

W klasie `Tape` dodaj metodę `clone` która będzie tworzyć nowy obiekt klasy `Tape` wraz z głęboką kopią listy z obiektu wywołującego. W tym przypadku będzie można wykorzystać Pythonową sztuckę do tworzenia kopii głębokiej, tj. `Tape(self.tape[:])`. W ten sposób nowy obiekt będzie posiadał kopię listy `self.tape`, a nie referencję, jak by było w przypadku wywołania `Tape(self.tape)`.

Następnie użyj tego wywołania w metodzie `reset`: `self.tapes = [tape.clone() for tape in self.initial_tapes]`.

Druga sprawa. Przenieś deklaracje zmiennych `self.tapes`, `self.state` oraz `self.match_position` do konstruktora `__init__`. Python niestety nie nakłada wymogów na deklarowanie zmiennych w klasie, ale dobrym zwyczajem jest deklarowanie wszystkich zmiennych w konstruktorze właśnie, tak żeby podczas analizowania struktury klasy można było szybko znaleźć wszystkie jej składowe w jednym miejscu ;)